src/AspNetCoreModuleV2/CommonLib/GlobalVersionUtility.cpp

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "stdafx.h"
+#include <experimental/filesystem>
+
+namespace fs = std::experimental::filesystem;
+
+// throws runtime error if no request handler versions are installed.
+// Throw invalid_argument if any argument is null
+std::wstring
+GlobalVersionUtility::GetGlobalRequestHandlerPath(PCWSTR pwzAspNetCoreFolderPath, PCWSTR pwzHandlerVersion, PCWSTR pwzHandlerName)
+{
+    if (pwzAspNetCoreFolderPath == NULL)
+    {
+        throw new std::invalid_argument("pwzAspNetCoreFolderPath is NULL");
+    }
+
+    if (pwzHandlerVersion == NULL)
+    {
+        throw new std::invalid_argument("pwzHandlerVersion is NULL");
+    }
+
+    if (pwzHandlerName == NULL)
+    {
+        throw new std::invalid_argument("pwzHandlerName is NULL");
+    }
+
+    std::wstring folderVersion(pwzHandlerVersion);
+    fs::path aspNetCoreFolderPath(pwzAspNetCoreFolderPath);
+
+    if (folderVersion.empty())
+    {
+        folderVersion = FindHighestGlobalVersion(pwzAspNetCoreFolderPath);
+    }
+
+    aspNetCoreFolderPath = aspNetCoreFolderPath
+        .append(folderVersion)
+        .append(pwzHandlerName);
+    return aspNetCoreFolderPath;
+}
+
+// Throw filesystem_error if directory_iterator can't iterate over the directory
+// Throw invalid_argument if any argument is null
+std::vector<fx_ver_t>
+GlobalVersionUtility::GetRequestHandlerVersions(PCWSTR pwzAspNetCoreFolderPath)
+{
+    if (pwzAspNetCoreFolderPath == NULL)
+    {
+        throw new std::invalid_argument("pwzAspNetCoreFolderPath is NULL");
+    }
+
+    std::vector<fx_ver_t> versionsInDirectory;
+    for (auto& p : fs::directory_iterator(pwzAspNetCoreFolderPath))
+    {
+        if (!fs::is_directory(p))
+        {
+            continue;
+        }
+
+        fx_ver_t requested_ver(-1, -1, -1);
+        if (fx_ver_t::parse(p.path().filename(), &requested_ver, false))
+        {
+            versionsInDirectory.push_back(requested_ver);
+        }
+    }
+    return versionsInDirectory;
+}
+
+// throws runtime error if no request handler versions are installed.
+// Throw invalid_argument if any argument is null
+std::wstring
+GlobalVersionUtility::FindHighestGlobalVersion(PCWSTR pwzAspNetCoreFolderPath)
+{
+    if (pwzAspNetCoreFolderPath == NULL)
+    {
+        throw new std::invalid_argument("pwzAspNetCoreFolderPath is NULL");
+    }
+
+    std::vector<fx_ver_t> versionsInDirectory = GetRequestHandlerVersions(pwzAspNetCoreFolderPath);
+    if (versionsInDirectory.empty())
+    {
+        throw new std::runtime_error("Cannot find request handler next to aspnetcorev2.dll. Verify a version of the request handler is installed in a version folder.");
```

Comment:
Resources?

Reply:
We need to discuss this. Exceptions only work with CHAR*, while resource and all other windows strings take WCHAR.